### PR TITLE
feat: add import detection (APP-77)

### PR DESCRIPTION
**Pull request by yaronskaya**

### CODE REVIEW

File: src/main/kotlin/app/extractors/CExtractor.kt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+class CExtractor : ExtractorInterface {
+    override fun extract(files: List<DiffFile>): List<CommitStats> {
+        TODO("not implemented")
+    }
+
+    override fun extractImports(fileContent: List<String>): List<String> {
+        val libraries = mutableSetOf<String>()
+
+        val quoteRegex = Regex("#import\\s+\"(\\w+)[/\\w+]*\\.\\w+\"")
```
Comment (anatolykislov) [inline src/main/kotlin/app/extractors/CExtractor.kt]: Did you test it? Probably shouldn't work.
Also, you can use only one regex with OR: `#import\s+[\"|<](\w+)[/\w+]*.\w+[\"|>]`

Comment (anatolykislov) [reply src/main/kotlin/app/extractors/CExtractor.kt]: Fix double slash everywhere.

File: src/main/kotlin/app/extractors/PhpExtractor.kt [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+class PhpExtractor : ExtractorInterface {
+    override fun extract(files: List<DiffFile>): List<CommitStats> {
+        TODO("not implemented")
+    }
+
+    override fun extractImports(fileContent: List<String>): List<String> {
+        val libraries = mutableSetOf<String>()
+
+        val regex = Regex("use\\s+(\\w+)[\\\\\\w+]*")
```
Comment (anatolykislov) [inline src/main/kotlin/app/extractors/PhpExtractor.kt]: require, require_once?

File: src/main/kotlin/app/extractors/CsharpExtractor.kt [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+class CsharpExtractor : ExtractorInterface {
+    override fun extract(files: List<DiffFile>): List<CommitStats> {
+        TODO("not implemented")
+    }
+
+    override fun extractImports(fileContent: List<String>): List<String> {
+        val libraries = mutableSetOf<String>()
+
+        val csLibraries = File("data/libraries/cs_libraries.txt")
```
Comment (anatolykislov) [inline src/main/kotlin/app/extractors/CsharpExtractor.kt]: This should be definitely optimized cuz everytime extractImports called a file will be read. IO operations take a lot of time. I'll do it.

File: src/main/kotlin/app/extractors/JavaExtractor.kt [hunk 4]
```diff
@@ -54,4 +56,28 @@ class JavaExtractor : ExtractorInterface {
 
         return stats
     }
+
+    override fun extractImports(fileContent: List<String>): List<String> {
+        val libraries = mutableSetOf<String>()
+
+        val javaLibraries = File("data/libraries/java_libraries.txt")
```
Comment (anatolykislov) [inline src/main/kotlin/app/extractors/JavaExtractor.kt]: In some Extractors you using list of imports, in some not, why?

File: src/main/kotlin/app/extractors/ObjectiveCExtractor.kt [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+class ObjectiveCExtractor : ExtractorInterface {
+    override fun extract(files: List<DiffFile>): List<CommitStats> {
+        TODO("not implemented")
+    }
+
+    override fun extractImports(fileContent: List<String>): List<String> {
+        val libraries = mutableSetOf<String>()
+
+        val importQuotesRegex = Regex("#import\\s+\"(\\w+)[/\\w+]*\\.\\w+\"")
```
Comment (anatolykislov) [inline src/main/kotlin/app/extractors/ObjectiveCExtractor.kt]: Use OR.

File: src/main/kotlin/app/extractors/CsharpExtractor.kt [hunk 6]
```diff
@@ -0,0 +1,38 @@
+// Copyright 2017 Sourcerer Inc. All Rights Reserved.
+// Author: Liubov Yaronskaya (lyaronskaya@sourcerer.io)
+
+package app.extractors
+
+import app.model.CommitStats
+import app.model.DiffFile
+import java.io.File
+
+class CsharpExtractor : ExtractorInterface {
```
Comment (anatolykislov) [inline src/main/kotlin/app/extractors/CsharpExtractor.kt]: CSharpExtractor

File: src/main/kotlin/app/extractors/CExtractor.kt [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+// Author: Liubov Yaronskaya (lyaronskaya@sourcerer.io)
+
+package app.extractors
+
+import app.model.CommitStats
+import app.model.DiffFile
+
+class CExtractor : ExtractorInterface {
+    override fun extract(files: List<DiffFile>): List<CommitStats> {
+        TODO("not implemented")
```
Comment (anatolykislov) [inline src/main/kotlin/app/extractors/CExtractor.kt]: This mean that exception will be thrown if this method will be called. Not a good idea to merge such things to master.

File: src/main/kotlin/app/extractors/CExtractor.kt (hunk 1 above)
Comment (yandzee) [reply src/main/kotlin/app/extractors/CExtractor.kt]: `#include` maybe?

File: src/main/kotlin/app/extractors/CppExtractor.kt [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+class CppExtractor : ExtractorInterface {
+    override fun extract(files: List<DiffFile>): List<CommitStats> {
+        TODO("not implemented")
+    }
+
+    override fun extractImports(fileContent: List<String>): List<String> {
+        val libraries = mutableSetOf<String>()
+
+        val quoteRegex = Regex("#import\\s+\"(\\w+)[/\\w+]*\\.\\w+\"")
```
Comment (yandzee) [inline src/main/kotlin/app/extractors/CppExtractor.kt]: Same

File: src/main/kotlin/app/extractors/JavaExtractor.kt [hunk 9]
```diff
@@ -54,4 +56,28 @@ class JavaExtractor : ExtractorInterface {
...
+            .inputStream().bufferedReader()
+            .readLines()
+            .toSet()
+
+        val regex = Regex("import\\s+(\\w+[.\\w+]*)")
+        fileContent.forEach {
+            val res = regex.find(it)
+            if (res != null) {
```
Comment (yandzee) [inline src/main/kotlin/app/extractors/JavaExtractor.kt]: `if (res == null) return` – inverse it to achieve lesser amount of nesting

File: src/main/kotlin/app/extractors/PhpExtractor.kt (hunk 2 above)
Comment (yandzee) [reply src/main/kotlin/app/extractors/PhpExtractor.kt]: include, include once?

File: src/main/kotlin/app/extractors/CExtractor.kt (hunk 1 above)
Comment (yaronskaya) [reply src/main/kotlin/app/extractors/CExtractor.kt]: Fixed.

File: src/main/kotlin/app/extractors/CsharpExtractor.kt (hunk 6 above)
Comment (yaronskaya) [reply src/main/kotlin/app/extractors/CsharpExtractor.kt]: .

File: src/main/kotlin/app/extractors/CsharpExtractor.kt (hunk 3 above)
Comment (yaronskaya) [reply src/main/kotlin/app/extractors/CsharpExtractor.kt]: .

File: src/main/kotlin/app/extractors/ObjectiveCExtractor.kt (hunk 5 above)
Comment (yaronskaya) [reply src/main/kotlin/app/extractors/ObjectiveCExtractor.kt]: Fixed.

File: src/main/kotlin/app/extractors/PhpExtractor.kt (hunk 2 above)
Comment (yaronskaya) [reply src/main/kotlin/app/extractors/PhpExtractor.kt]: Fixed.

File: src/main/kotlin/app/extractors/JavaExtractor.kt (hunk 4 above)
Comment (yaronskaya) [reply src/main/kotlin/app/extractors/JavaExtractor.kt]: It depends. In js, there are sometimes no import lines, that's why we tokenize all lines and see what tokens are in the import list. In this case we should catch things like ((com|ru).)?organization.library.modules.* and also skip imports of inner files.

Comment (anatolykislov) [reply src/main/kotlin/app/extractors/JavaExtractor.kt]: So you gonna use this libraries files in a classifier for every language?

File: src/main/kotlin/app/model/DiffEdit.kt [hunk 10]
```diff
@@ -6,13 +6,11 @@ package app.model
 import org.eclipse.jgit.diff.Edit
```
Comment (sergey-surkov) [inline src/main/kotlin/app/model/DiffEdit.kt]: copyright/author

File: src/main/kotlin/app/extractors/ExtractorInterface.kt [hunk 11]
```diff
@@ -1,11 +1,32 @@
...
 import app.model.DiffFile
 import app.model.CommitStats
 
 interface ExtractorInterface {
-    fun extract(files: List<DiffFile>): List<CommitStats>
+    fun extract(files: List<DiffFile>): List<CommitStats> {
+        files.map { file ->
+            file.old.imports = extractImports(file.old.content)
+            file.new.imports = extractImports(file.new.content)
+            file
```
[the remark this answers is not in the document]
Comment (anatolykislov) [reply src/main/kotlin/app/extractors/ExtractorInterface.kt]: Last line is return value.

File: src/main/kotlin/app/extractors/Extractor.kt [hunk 12]
```diff
@@ -1,45 +1,32 @@
...
 import app.model.CommitStats
 import app.model.DiffFile
 
 class Extractor : ExtractorInterface {
-    companion object Constants {
+    companion object {
         val TYPE_LANGUAGE = 1
         val TYPE_KEYWORD = 2
-
         val SEPARATOR = ">"
```
[the remark this answers is not in the document]
Comment (anatolykislov) [reply src/main/kotlin/app/extractors/Extractor.kt]: Used in JavaExtractor for example to separate language from tech (or keywords, etc). "Java>Keyword".

File: src/main/kotlin/app/model/DiffEdit.kt (hunk 10 above)
Comment (anatolykislov) [reply src/main/kotlin/app/model/DiffEdit.kt]: There is.

**Pull request merged**